src/cipher.rs

```diff
@@ -33,7 +33,8 @@ pub struct rustls_certificate {
     _private: [u8; 0],
 }
 
-impl CastPtr for rustls_certificate {
+impl Castable for rustls_certificate {
```

Comment:
This type and `rustls_supported_ciphersuite` are unique in that unlike most of the other snake types they only implemented `CastPtr` and not one of `BoxCastPtr` or `ArcCastPtr`.

I think using `ArcCastPtr` makes sense here since the new `Castable` type requires choosing a `CastType`. It didn't feel like breaking out a separate representation for this case was worthwhile, but its possible I'm overlooking something that would require more care here.

Reply:
Currently `rustls_certificate` is only given to the user in two places:

 - https://github.com/rustls/rustls-ffi/blob/8caf0c8e2c239251f3d3d2f307ede461af50e984/src/rustls.h#L913-L921
 - https://github.com/rustls/rustls-ffi/blob/8caf0c8e2c239251f3d3d2f307ede461af50e984/src/rustls.h#L1355-L1368

In both of these cases, the underlying Rust type is a reference, not a Box or an Arc. I think what we need here is a new type, `RefCastPtr`, that expresses a non-owned reference.

If in the future we provide any methods that give the C code an owned pointer to a `rustls::Certificate` (a `Box<Certificate>` or an `Arc<Certificate>`), we would want to provide a separate C type (`rustls_owned_certificate` or `rustls_certificate_owned`) to prevent any confusion.

Reply:
> In both of these cases, the underlying Rust type is a reference, not a Box or an Arc.

Ahh, ok that makes sense, and I think I understand why nothing blew up when I had this arranged incorrectly based on how the types are used.

>  I think what we need here is a new type, RefCastPtr, that expresses a non-owned reference.

I used the suggestion from the other comment thread for a `OwnershipRef` to use as the `Ownership` associated type in the `Castable` impl for these types. 

> This type and rustls_supported_ciphersuite are unique

There was actually a third instance too (also now fixed up): `rustls_client_hello`. All three now implement `Castable` with `Ownership = OwnershipRef`.

Since `try_ref_from_ptr` and the underlying `try_from` are generic over the `Ownership` of the `Castable` we can use `try_ref_from_ptr` for these types but not the arc or box specific variants. I think that's what we want :+1: